src/main/java/net/floodlightcontroller/forwarding/Forwarding.java

```diff
@@ -124,21 +130,117 @@ public Command processPacketInMessage(IOFSwitch sw, OFPacketIn pi, IRoutingDecis
 
         return Command.CONTINUE;
     }
+    
+    /**
+	 * Builds a cookie that includes routing decision information.
+	 *
+	 * @param decision The routing decision providing a descriptor, or null
+	 * @return A cookie with our app id and the required routing fields masked-in
+	 */
+	protected U64 makeForwardingCookie(IRoutingDecision decision) {
+		int user_fields = 0;
+
+		U64 decision_cookie = (decision == null) ? null : decision.getDescriptor();
+		if (decision_cookie != null) {
+			user_fields |= AppCookie.extractUser(decision_cookie) & DECISION_MASK;
+		}
+
+		// TODO: Mask in any other required fields here (e.g. link failure flowset ID)
+
+		if (user_fields == 0) {
+			return DEFAULT_FORWARDING_COOKIE;
+		}
+		return AppCookie.makeCookie(FORWARDING_APP_ID, user_fields);
+	}
+
+	/** Called when the handleDecisionChange is triggered by an event (routing decision was changed in firewall).
+	 *  
+	 *  @param changedDecisions Masked routing descriptors for flows that should be deleted from the switch.
+	 */
+	@Override
+	public void routingDecisionChanged(Iterable<Masked<U64>> changedDecisions) {
+		deleteFlowsByDescriptor(changedDecisions);
+	}
+	
+	/**
+	 * Converts a sequence of masked IRoutingDecision descriptors into masked Forwarding cookies.
+	 *
+	 * This generates a list of masked cookies that can then be matched in flow-mod messages.
+	 *
+	 * @param maskedDescriptors A sequence of masked cookies describing IRoutingDecision descriptors
+	 * @return A collection of masked cookies suitable for flow-mod operations
+	 */
+	protected Collection<Masked<U64>> convertRoutingDecisionDescriptors(Iterable<Masked<U64>> maskedDescriptors) {
+		if(maskedDescriptors == null){
+			return null;
+		}
+		List<Masked<U64>> result = new ArrayList<Masked<U64>>();
+
+		for (Masked<U64> maskedDescriptor : maskedDescriptors) {
+			long user_mask = AppCookie.extractUser(maskedDescriptor.getMask()) & DECISION_MASK;
+			long user_value = AppCookie.extractUser(maskedDescriptor.getValue()) & user_mask;
+
+			// TODO combine in any other cookie fields you need here.
+
+			result.add(
+					Masked.of(
+							AppCookie.makeCookie(FORWARDING_APP_ID, (int)user_value),
+							AppCookie.getAppFieldMask().or(U64.of(user_mask))
+					)
+			);
+		}
+
+		return result;
+	}
+
+	/**
+	 * On all active switches, deletes all flows matching the IRoutingDecision descriptors provided
+	 * as arguments.
+	 *
+	 * @param descriptors The descriptors and masks describing which flows to delete.
+	 */
+	protected void deleteFlowsByDescriptor(Iterable<Masked<U64>> descriptors) {
+		Collection<Masked<U64>> masked_cookies = convertRoutingDecisionDescriptors(descriptors);
+
+		if (masked_cookies != null && !masked_cookies.isEmpty()) {
+			for (DatapathId dpid : switchService.getAllSwitchDpids()) {
+				IOFSwitch sw = switchService.getActiveSwitch(dpid);
+				if (sw == null) {
+					continue;
+				}
+
+				// Would like to swap these for loops and only build the message set once,
```

Comment:
Very true. To help reduce having to build all the flow-delete messages for each switch, you could cache the messages per OpenFlow version. This will guarantee that the list of messages will only be constructed once per OpenFlow version. Any subsequent switches that happen to use the same OpenFlow version can reuse the already-built list of messages.